LISTENER_MIGRATION.md

```diff
@@ -0,0 +1,338 @@
+        ...
+        googlePayClient.requestPayment(requireActivity(), googlePayRequest)
+    }
+
+    override fun onGooglePaySuccess(paymentMethodNonce: PaymentMethodNonce) {
+        // send nonce to server and create a transaction
+    }
+
+    override fun onGooglePayFailure(error: Exception) {
+        // handle error
+
+    }
+}
+```
+
+## Venmo
+
+```kotlin
+// MerchantActivity.kt
+class MerchantActivity : AppCompatActivity(), VenmoListener {
+    
+    lateinit var braintreeClient: BraintreeClient
+    lateinit var venmoClient: VenmoClient
+    
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        
+        braintreeClient =
+          BraintreeClient(this, MerchantClientTokenProvider())
+          
+        venmoClient = VenmoClient(this, braintreeClient)
+        venmoClient.listener = this
+    }
+    
+    fun launchVenmo() {
+        ...
+        venmoClient.tokenizeVenmoAccount(this, venmoRequest)
+    }
+   
+    override fun onVenmoSuccess(venmoAccountNonce: VenmoAccountNonce) {
+        // send nonce to server and create a transaction
+    }
+    
+    override fun onVenmoFailure(error: Exception) {
+        // handle error
+    }
+}
+
+
+// MerchantFragment.kt
+class MerchantFragment: Fragment(), VenmoListener {
+    
+    lateinit var braintreeClient: BraintreeClient
+    lateinit var venmoClient: VenmoClient
+    
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        
+        braintreeClient =
+          BraintreeClient(this, MerchantClientTokenProvider())
+          
+        venmoClient = VenmoClient(this, braintreeClient)
+        venmoClient.listener = this
+    }
+
+    fun launchVenmo() {
+        ...
+        venmoClient.tokenizeVenmoAccount(requireActivity(), venmoRequest)
+    }
+    
+    override fun onVenmoSuccess(venmoAccountNonce: venmoAccountNonce) {
+        // send nonce to server and create a transaction
+    }
+    
+    override fun onVenmoFailure(error: Exception) {
+        // handle error
+    }
+}
+```
+
+## ThreeDSecure
+
+```kotlin
+// MerchantActivity.kt
+class MerchantActivity : AppCompatActivity(), ThreeDSecureListener {
+    
+    lateinit var braintreeClient: BraintreeClient
+    lateinit var threeDSecureClient: ThreeDSecureClient
+    
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        
+        braintreeClient =
+          BraintreeClient(this, MerchantClientTokenProvider())
+          
+        threeDSecureClient = ThreeDSecureClient(this, braintreeClient)
+        threeDSecureClient.listener = this
+    }
+    
+    private fun launch3DS() {
+        ...
+        threeDSecureClient.performVerification(this, threeDSecureRequest) { threeDSecureLookupResult, lookupError ->
+          threeDSecureClient.continuePerformVerification(this@MerchantActivity, threeDSecureRequest, threeDSecureLookupResult)
```

Comment:
future: I wonder if we can convert this perform verification callback into additional listener methods?

Reply:
Could be next major version too when we make this the primary integration method.

Reply:
Yeah I think we did the `performVerification` with a callback because of the server side 3DS option - if we added another listener method, merchants not needing it would still have to implement it.

Reply:
Yeah that's true I forgot about that part of it.